test/e2e/llmisvc/test_llm_inference_service.py

```diff
@@ -0,0 +1,324 @@
+        
+    except Exception as e:
+        raise ValueError(f"Failed to get URL for LLM inference service {service_name}: {e}")
+    
+def llm_isvc_watch(name=None, namespace=None, timeout_seconds=600, generation=0):
+    """
+    Watch LLM inference service until all conditions are ready
+    """
+    headers = ["NAME", "READY", "WORKLOADS", "ROUTER", "MAIN", "URL"]
+    table_fmt = "plain"
+
+    stream = k8s_watch.Watch().stream(
+        client.CustomObjectsApi().list_namespaced_custom_object,
+        KSERVE_GROUP,
+        KSERVE_V1ALPHA1_VERSION,
+        namespace,
+        KSERVE_PLURAL_LLMINFERENCESERVICE,
+        timeout_seconds=timeout_seconds,
+    )
+
+    for event in stream:
+        llm_isvc = event["object"]
+        llm_isvc_name = llm_isvc["metadata"]["name"]
+        if name and name != llm_isvc_name:
+            continue
+        else:
+            ready = "Unknown"
+            workloads_ready = "Unknown"
+            router_ready = "Unknown"
+            main_workload_ready = "Unknown"
+            url = ""
+            
+            if llm_isvc.get("status", ""):
+                status = llm_isvc["status"]
+                
+                if "url" in status and status["url"]:
+                    url = status["url"]
+                elif "addresses" in status and status["addresses"]:
+                    first_address = status["addresses"][0]
+                    if "url" in first_address:
+                        url = first_address["url"]
+                
+                if generation != 0:
+                    observed_generation = status.get("observedGeneration")
+                    if observed_generation != generation:
+                        continue
+                
+                conditions_status = {}
+                for condition in status.get("conditions", []):
+                    condition_type = condition.get("type", "")
+                    condition_status = condition.get("status", "Unknown")
+                    conditions_status[condition_type] = condition_status
+                
+                ready = conditions_status.get("Ready", "Unknown")
+                workloads_ready = conditions_status.get("WorkloadsReady", "Unknown")
+                router_ready = conditions_status.get("RouterReady", "Unknown")
+                main_workload_ready = conditions_status.get("MainWorkloadReady", "Unknown")
+                
+                print(
+                    tabulate(
+                        [[llm_isvc_name, ready, workloads_ready, router_ready, main_workload_ready, url]],
+                        headers=headers,
+                        tablefmt=table_fmt,
+                    )
+                )
+                
+                if (ready == "True" and 
+                    workloads_ready == "True" and 
+                    router_ready == "True" and
+                    main_workload_ready == "True"):
+                    print("All conditions ready")
```

Comment:
I noticed it fails when a request is made immediately after it becomes ready. To improve the stability of the end-to-end tests, I would suggest adding a 1 or 2-second sleep here.